src/vsr/replica.zig

```diff
@@ -3044,6 +3044,7 @@ pub fn ReplicaType(
 
             self.pulse_timeout.reset();
             if (!self.pulse_needed()) return;
+            if (self.pipeline.queue.full()) return;
```

Comment:
If you swap the order of these two checks:
```suggestion
            if (self.pipeline.queue.full()) return;
            if (!self.pulse_needed()) return;
```

Then `pulse_needed` could assert that `!self.pipeline.queue.full()`. Which in turn "documents" the reason that this `full()` check is here, rather than inside `pulse_needed` -- `pulse_needed()`s other call sites are guaranteed to have a non-full pipeline.